File: edu.wpi.first.wpilib.plugins.cpp/resources/templates/examples/PacGoat/src/Commands/DriveForward.cpp
#include "DriveForward.h"

#include "Robot.h"

/*
 * All WPILib classes are in the FRC namespace. Either `using namespace frc` 
 * or frc scope (ex. `frc::RobotDrive`) are required to use WPILib functionality. 
 * For headers, use scope instead of `using namespace frc` to avoid global namespace
 * pollution. For source files, either option is viable.
 * See http://wpilib.screenstepslive.com/s/4485/m/13810/l/680719 for more information
 */
using namespace frc;

void DriveForward::init(double dist, double maxSpeed) {
	Requires(Robot::drivetrain.get());
	distance = dist;
	driveForwardSpeed = maxSpeed;
	error = 0;
}

DriveForward::DriveForward() {
	init(10, 0.5);
}

DriveForward::DriveForward(double dist) {
	init(dist, 0.5);
}

DriveForward::DriveForward(double dist, double maxSpeed) {
	init(dist, maxSpeed);
}

// Called just before this Command runs the first time
void DriveForward::Initialize() {
	Robot::drivetrain->GetRightEncoder()->Reset();
	SetTimeout(2);
}

// Called repeatedly when this Command is scheduled to run
void DriveForward::Execute() {
	error = (distance - Robot::drivetrain->GetRightEncoder()->GetDistance());
	if (driveForwardSpeed * KP * error >= driveForwardSpeed) {
		Robot::drivetrain->TankDrive(driveForwardSpeed, driveForwardSpeed);
	} else {
		Robot::drivetrain->TankDrive(driveForwardSpeed * KP * error,
				driveForwardSpeed * KP * error);
	}
}

// Make this return true when this Command no longer needs to run execute()
bool DriveForward::IsFinished() {
	return (abs(error) <= TOLERANCE) || IsTimedOut();
}

// Called once after isFinished returns true
void DriveForward::End() {
	Robot::drivetrain->Stop();
}

// Called when another command which requires one or more of the same
// subsystems is scheduled to run
void DriveForward::Interrupted() {
	End();
}

File: edu.wpi.first.wpilib.plugins.java/resources/templates/examples/PacGoat/src/org/usfirst/frc/team190/pacgoat/commands/DriveForward.java
package $package.commands;

import edu.wpi.first.wpilibj.command.Command;

import $package.Robot;

/**
 * This command drives the robot over a given distance with simple proportional
 * control This command will drive a given distance limiting to a maximum speed.
 */
public class DriveForward extends Command {
	private double driveForwardSpeed;
	private double distance;
	private double error;
	private final double kTolerance = 0.1;
	private final double kP = -1.0 / 5.0;

	public DriveForward() {
		this(10, 0.5);
	}

	public DriveForward(double dist) {
		this(dist, 0.5);
	}

	public DriveForward(double dist, double maxSpeed) {
		requires(Robot.drivetrain);
		distance = dist;
		driveForwardSpeed = maxSpeed;
	}

	@Override
	protected void initialize() {
		Robot.drivetrain.getRightEncoder().reset();
		setTimeout(2);
	}

	@Override
	protected void execute() {
		error = (distance - Robot.drivetrain.getRightEncoder().getDistance());
		if (driveForwardSpeed * kP * error >= driveForwardSpeed) {
			Robot.drivetrain.tankDrive(driveForwardSpeed, driveForwardSpeed);
		} else {
			Robot.drivetrain.tankDrive(driveForwardSpeed * kP * error, driveForwardSpeed * kP * error);
		}
	}

	@Override
	protected boolean isFinished() {
		return (Math.abs(error) <= kTolerance) || isTimedOut();
	}

	@Override
	protected void end() {
		Robot.drivetrain.stop();
	}
}
